core/state-machine/src/overlayed_changes.rs

```diff
@@ -74,17 +408,158 @@ impl OverlayedChangeSet {
 		self.top.is_empty() && self.children.is_empty()
 	}
 
-	/// Clear the change set.
-	pub fn clear(&mut self) {
-		self.top.clear();
-		self.children.clear();
+	/// Discard prospective changes to state.
+	pub fn discard_prospective(&mut self) {
+		let mut i = self.history.len();
+		while i > 0 {
+			i -= 1;
+			match self.history[i] {
```

Comment:
This would break the property of this cache. I will write a gc method for cleaning (get_mut does only clean the simple case without going to deep as we can expect many get_mut call between dropped).
I tend to agree that memory gc could be triggered on 'drop_perspective' and 'commit_perspectve' : those operation being not frequent.
Gc on those operation could also be conditional to actual used (estimated) memory: we may want to chain small extrinsic operation without memory mgmt.

Truncating history can also be done up to the last commit after a full gc (we will need a offset).